setup.py

```diff
@@ -1,231 +1,261 @@
-        package_data=PACKAGE_DATA,
-        ext_modules=exts,
-        cmdclass={'build_ext': build_ext_cython_subclass},
-        license="MIT",
-    )
+    include_dirs = [
+        get_python_inc(plat_specific=True),
+        os.path.join(src_path, 'include')]
 
+    ext_modules = []
+    for mod_name in MOD_NAMES:
+        mod_path = mod_name.replace('.', '/') + '.cpp'
+        ext_modules.append(
+            Extension(mod_name, [mod_path],
+                language='c++', include_dirs=include_dirs))
 
-def run_setup(exts):
-    setup(
+    metadata = dict(
         name='spacy',
-        packages=['spacy', 'spacy.tokens', 'spacy.en', 'spacy.serialize',
-                  'spacy.syntax', 'spacy.munge',
-                  'spacy.tests',
-                  'spacy.tests.matcher',
-                  'spacy.tests.morphology',
-                  'spacy.tests.munge',
-                  'spacy.tests.parser',
-                  'spacy.tests.serialize',
-                  'spacy.tests.spans',
-                  'spacy.tests.tagger',
-                  'spacy.tests.tokenizer',
-                  'spacy.tests.tokens',
-                  'spacy.tests.vectors',
-                  'spacy.tests.vocab'],
-        description="Industrial-strength NLP",
+        packages=PACKAGES,
+        description='Industrial-strength NLP',
         author='Matthew Honnibal',
-        author_email='honnibal@gmail.com',
+        author_email='matt@spacy.io',
         version=VERSION,
-        url="http://honnibal.github.io/spaCy/",
-        package_data=PACKAGE_DATA,
-        ext_modules=exts,
-        license="MIT",
+        url='https://spacy.io',
+        license='MIT',
+        ext_modules=ext_modules,
         install_requires=['numpy', 'murmurhash == 0.24', 'cymem == 1.30', 'preshed == 0.44',
-                          'thinc == 4.0.0', "text_unidecode", 'plac', 'six',
-                          'ujson', 'cloudpickle', 'sputnik == 0.5.2'],
-        setup_requires=["headers_workaround"],
-        cmdclass = {'build_ext': build_ext_subclass },
+                          'thinc == 4.0.0', 'text_unidecode', 'plac', 'six',
+                          'ujson', 'cloudpickle', 'sputnik == 0.6.2'],
+        cmdclass = {
+            'build_ext': build_ext_subclass},
     )
 
-    import headers_workaround
+    # Run build
+    cwd = os.path.abspath(os.path.dirname(__file__))
+    if not os.path.exists(os.path.join(cwd, 'PKG-INFO')):
+        # Generate Cython sources, unless building from source release
+        generate_cython()
 
-    headers_workaround.fix_venv_pypy_include()
-    headers_workaround.install_headers('murmurhash')
-    headers_workaround.install_headers('numpy')
+        # sync include dirs from native dependencies
+        include_dir = os.path.join(src_path, 'include')
+        if os.path.exists(include_dir):
+            shutil.rmtree(include_dir)
+        os.mkdir(include_dir)
 
+        import numpy
+        shutil.copytree(
```

Comment:
Random paranoia:

* How does this handle missing directories? Are these paths ever missing?
* How does this handle symlinks? Are these paths ever symlinks?
* How does this handle permissions errors? Do we ever have permission errors here?